### PR TITLE
fix(DnD): [#FOR-531] fix targetting system on drag and drop

### PR DESCRIPTION
**Pull request by qperie**
## Describe your changes
The previously implemented targetting system is know adapting when form elements are drag and drop from editor view or organization pop up

## Checklist tests
Play with drag and drop in editor view or organization pop up  and check that targetting values are still logic after each move

## Issue ticket number and link
FOR-531 : https://jira.support-ent.fr/browse/FOR-531

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: common/src/main/resources/ts/models/FormElement/FormElement.ts [hunk 1]
```diff
@@ -49,6 +49,7 @@ export abstract class FormElement implements Selectable {
     }
 
     equals = (formElement: FormElement) : boolean => {
+        if (!formElement) return false;
         return this.form_element_type === formElement.form_element_type && this.id === formElement.id;
```
Comment (BastienBenard) [inline common/src/main/resources/ts/models/FormElement/FormElement.ts:53]: return this.form_element_type === formElement.form_element_type && this.id === formElement.id;
```suggestion
        return formElement && this.form_element_type === formElement.form_element_type && this.id === formElement.id;
```

File: common/src/main/resources/ts/models/FormElement/FormElement.ts [hunk 2]
```diff
@@ -61,15 +62,16 @@ export abstract class FormElement implements Selectable {
     getFollowingFormElement = (formElements: FormElements) : FormElement => {
         // Case formElement is not formElement but question inside a section
         if (this instanceof Question && this.section_id) {
-            let parent: Section = this.getParentSection(formElements);
-            let nextElements: FormElement[] = formElements.all.filter((e: FormElement) => e.id === parent.position + 1);
-            return nextElements.length == 1 ? nextElements[0] : null;
+            let parentSection: Section = this.getParentSection(formElements);
+            let followingPosition: number = parentSection.position + 1;
+            let followingElements: FormElement[] = formElements.all.filter((e: FormElement) => e.position === followingPosition);
+            return followingElements.length == 1 ? followingElements[0] : null;
```
Comment (BastienBenard) [inline common/src/main/resources/ts/models/FormElement/FormElement.ts:68]:
```suggestion
            return formElements.all.find((e: FormElement) => e.position === followingPosition);
```

File: common/src/main/resources/ts/models/FormElement/FormElement.ts [hunk 3]
```diff
@@ -61,15 +62,16 @@ export abstract class FormElement implements Selectable {
...
         }
 
         // Case formElement is section without target or just a solo question
-        let nextPosition: number = this.position + 1;
-        let nextElements: FormElement[] = formElements.all.filter((e: FormElement) => e.position === nextPosition);
-        return nextElements.length == 1 ? nextElements[0] : null;
+        let followingPosition: number = this.position + 1;
+        let followingElements: FormElement[] = formElements.all.filter((e: FormElement) => e.position === followingPosition);
+        return followingElements.length == 1 ? followingElements[0] : null;
```
Comment (BastienBenard) [inline common/src/main/resources/ts/models/FormElement/FormElement.ts:74]: same

File: formulaire/src/main/java/fr/openent/formulaire/controllers/FormElementController.java [hunk 4]
```diff
@@ -91,14 +103,22 @@ public void update(HttpServerRequest request) {
...
                 renderError(request);
                 return;
             }
 
             List<FormElement> formElements = FormElement.toListFormElements(formElementsJson);
             formElementService.update(formElements, formId)
+                .compose(updatedFormElements -> {
+                    List<QuestionChoice> choices = formElements.stream()
+                            .filter(e -> e instanceof Question)
+                            .flatMap(e -> ((Question) e).getChoices().stream())
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/FormElementController.java:117]:
```suggestion
                            .map(Question.class::cast)
                            .flatMap(question -> question.getChoices().stream())
```

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultQuestionChoiceService.java [hunk 5]
```diff
@@ -111,6 +120,41 @@ public Future<JsonObject> update(QuestionChoice choice, String locale) {
         return promise.future();
     }
 
+    @Override
+    public Future<JsonArray> update(List<QuestionChoice> choices, String locale) {
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultQuestionChoiceService.java:124]: Tu peux essayer de faire un model et de retourner future de liste de model a la place d'un JsonArray

File: formulaire/src/test/java/fr/openent/formulaire/service/test/impl/DefaultQuestionChoiceServiceTest.java [hunk 6]
```diff
@@ -65,7 +72,98 @@ public void testCreate(TestContext ctx) {
...
+//            ctx.assertEquals(TRANSACTION, body.getString(ACTION));
+//            ctx.assertEquals(expectedQuery, body.getJsonArray(STATEMENTS).toString());
+//            async.complete();
+//        });
+//
+//        defaultFormElementService.update(formElements, "1")
+//                .onSuccess(result -> async.complete());
+//
+//        async.awaitSuccess(10000);
+
```
Comment (BastienBenard) [inline formulaire/src/test/java/fr/openent/formulaire/service/test/impl/DefaultQuestionChoiceServiceTest.java:166]: Pourquoi ?

File: formulaire/src/main/java/fr/openent/formulaire/controllers/QuestionChoiceController.java [hunk 7]
```diff
@@ -24,6 +24,7 @@
 import static fr.openent.form.core.constants.ShareRights.CONTRIB_RESOURCE_RIGHT;
 import static fr.openent.form.core.enums.ApiVersions.*;
 import static fr.openent.form.helpers.RenderHelper.renderInternalError;
+import static org.entcore.common.http.response.DefaultResponseHandler.arrayResponseHandler;
```
Comment (flmartineau) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/QuestionChoiceController.java:27]: Tu t'en sers de cet import ?

**Review by flmartineau — APPROVED** [review]

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultQuestionChoiceService.java (hunk 5 above)
Comment (qperie) [reply formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultQuestionChoiceService.java:124]: J'comprends l'idée mais à l'heure actuelle le seul endroit utilisé on a besoin d'un JsonArray derrière donc ça me paraît overkill d'ajouter 200 lignes de helpers pour typer un JsonArray en List<QuestionChoice> pour le retransformer en JsonArray juste après non ?

File: formulaire/src/test/java/fr/openent/formulaire/service/test/impl/DefaultQuestionChoiceServiceTest.java (hunk 6 above)
Comment (qperie) [reply formulaire/src/test/java/fr/openent/formulaire/service/test/impl/DefaultQuestionChoiceServiceTest.java:166]: Parce que je faisais plusieurs trucs en même temps et j'ai oublié d'enlever ça x)

File: formulaire/src/main/java/fr/openent/formulaire/controllers/QuestionChoiceController.java (hunk 7 above)
Comment (qperie) [reply formulaire/src/main/java/fr/openent/formulaire/controllers/QuestionChoiceController.java:27]: Nope, bien vu

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultQuestionChoiceService.java (hunk 5 above)
Comment (qperie) [reply formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultQuestionChoiceService.java:121]: Du coup vu le timing j'ai laissé tel quel et mis de coté de quoi implémenter ça au prochain sprint 👍

**Review by BastienBenard — APPROVED** [review]

**Pull request merged**